spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/quantized/Linear.scala

```diff
@@ -0,0 +1,203 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.intel.analytics.bigdl.nn.quantized
+
+import com.intel.analytics.bigdl.bigquant.BigQuant
+import com.intel.analytics.bigdl.nn.ErrorInfo
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+import com.intel.analytics.bigdl.tensor._
+import com.intel.analytics.bigdl.utils.serializer.{DataConverter, ModuleData, ModuleSerializer}
+import com.intel.analytics.bigdl.utils.{T, Table}
+import java.io.{IOException, ObjectInputStream, ObjectOutputStream}
+import scala.reflect.ClassTag
+import scala.reflect.runtime.universe
+import serialization.Bigdl.{AttrValue, BigDLModule}
+
+private[bigdl] class Linear[T: ClassTag](
+  val inputSize: Int,
+  val outputSize: Int,
+  val withBias: Boolean = true
+)(implicit ev: TensorNumeric[T]) extends QuantizedModule[T](outputSize) {
+
+  val data: QuantizedTensor[T] = QuantizedTensor[T]()
+  @transient var weight: QuantizedTensor[T] = _
+  val bias: Tensor[T] = Tensor[T](outputSize)
+
+  private def initWeightAndBias(weightFP32: Tensor[T], biasFP32: Tensor[T]): this.type = {
+    if (biasFP32 != null) {
+      bias.copy(biasFP32)
+    } else {
+      bias.fill(ev.fromType(0)) // TODO bias may be null, at that time, we should not initialize it
+    }
+
+    val weightFP32Tmp = weightFP32.view(Array(outputSize, inputSize))
+    val params = LinearWeightParams(outputSize, inputSize)
+    weight = QuantizedTensor[T](weightFP32Tmp, params, LinearWeight)
+
+    this
+  }
+
+  override def init(): this.type = {
+    weight.init(LinearWeightParams(outputSize, inputSize), LinearWeight)
+
+    this
+  }
+
+  @throws(classOf[IOException])
+  private def writeObject(out: ObjectOutputStream): Unit = {
+    out.defaultWriteObject()
+
+    out.writeObject(weight)
+  }
+
+  @throws(classOf[IOException])
+  private def readObject(in: ObjectInputStream): Unit = {
+    in.defaultReadObject()
+
+    weight = in.readObject().asInstanceOf[QuantizedTensor[T]]
+
+    if (weight.getStorage != null && weight.getNativeStorage == 0L) {
+      init()
+    }
+  }
+
+  override def updateOutput(input: Tensor[T]): Tensor[T] = {
+    require(input.dim() == 1 || input.dim() == 2,
+      "bigquant.Linear: " + ErrorInfo.constrainInputAsVectorOrBatch)
+
+    val batchSize = if (input.dim() == 1) {
+      output.resize(Array(outputSize)) // TODO
+      1
+    } else {
+      output.resize(Array(input.size(1), outputSize))
+      require(inputSize == input.size(2), s"dimension error")
+      input.size(1)
+    }
+
+    data.init(LinearDataParams(batchSize, inputSize), LinearData)
```

Comment:
I think you can just check if the current data shape is different, and if so you can release it and create a new one.